source/slang/parser.cpp

```diff
@@ -598,10 +646,57 @@ namespace Slang
         AddModifier(&modifierLink, modifier);
     }
 
+    // 
+    // '::'? identifier ('::' identifier)* 
+    static Token parseAttributeName(Parser* parser)
+    {
+        const SourceLoc scopedIdSourceLoc = parser->tokenReader.PeekLoc();
+
+        // Strip initial :: if there is one
+        const TokenType initialTokenType = parser->tokenReader.PeekTokenType();
+        if (initialTokenType == TokenType::Scope)
+        {
+            parser->ReadToken(TokenType::Scope); 
+        }
+
+        const Token firstIdentifier = parser->ReadToken(TokenType::Identifier);
+        if (initialTokenType != TokenType::Scope && parser->tokenReader.PeekTokenType() != TokenType::Scope)
+        {
+            return firstIdentifier;
+        }
+
+        // Build up scoped string
+        StringBuilder scopedIdentifierBuilder;
+        if (initialTokenType == TokenType::Scope)
+        {
+            scopedIdentifierBuilder.Append('_'); 
+        }
+        scopedIdentifierBuilder.Append(firstIdentifier.Content);
+
+        while (parser->tokenReader.PeekTokenType() == TokenType::Scope)
+        {
+            parser->ReadToken(TokenType::Scope);
+            scopedIdentifierBuilder.Append('_'); 
+            
+            const Token nextIdentifier(parser->ReadToken(TokenType::Identifier));
+            scopedIdentifierBuilder.Append(nextIdentifier.Content);
+        }
+
+        // Make a 'token'
+        const String scopedIdentifier(scopedIdentifierBuilder.ToString());
+        Token token(TokenType::Identifier, scopedIdentifier, scopedIdSourceLoc);
+        token.ptrValue = getName(parser, token.Content);
+
+        return token;
+    }
+
     // Parse HLSL-style `[name(arg, ...)]` style "attribute" modifiers
     static void ParseSquareBracketAttributes(Parser* parser, RefPtr<Modifier>** ioModifierLink)
     {
         parser->ReadToken(TokenType::LBracket);
+
+        const bool hasDoubleBracket = AdvanceIf(parser, TokenType::LBracket);
```

Comment:
Oh, never mind. You used `AdvanceIf` here. Maybe there was a reason why it didn't make sense in the other function.

Reply:
I believe it was because I didn't want it to advance. It has somewhat overcomplicated logic - because I was trying to make the case just parsing a identifier - the common case not have to construct/destruct anything extra.